{% load static %}
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  <link rel="shortcut icon" type="image/png" href="{% static 'favicon.ico' %}"/>
  <link rel="stylesheet" type="text/css" href="{% static 'css/main.css' %}">
  <link rel="stylesheet" href="{% static 'css/bootstrap.min.css' %}">
  <title>GRAND</title>
</head>
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=GA_MEASUREMENT_ID"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-151168345-1');
</script>
<body class="d-flex flex-column min-vh-100">
<div class="wrapper flex-grow-1">

{% load widget_tweaks %}

<!-- Header -->
<nav class="navbar bg-dark navbar-dark navbar-expand-sm">
    <div class="container">
        <a class="navbar-brand m-0" href="/">
            <img src="{% static 'images/logo.svg' %}" style="width: 400px;">
        </a>
        <div class="navbar-nav">
            <a class="nav-item nav-link active" href="/">Home</a>
            <li class="nav-item dropdown">
            <a class="nav-link dropdown-toggle" data-toggle="dropdown" href="#">Networks</a>
            <div class="dropdown-menu">
              <a href="/drugs" class="dropdown-item">Small molecules</a>
              <a href="/cell" class="dropdown-item">Cell lines</a>
              <a href="/tissues" class="dropdown-item">Tissues</a>
            </div>
          </li>
             <li class="nav-item dropdown">
          <a class="nav-link dropdown-toggle" data-toggle="dropdown" href="#">Analysis</a>
          <div class="dropdown-menu">
            <a href="/analysis" class="dropdown-item">CLUEreg</a>
            <a href="/disease" class="dropdown-item">Disease/Tissues</a>
          </div>
        </li>
            <a class="nav-item nav-link" href="#">Help</a>
            <a class="nav-item nav-link" href="/about">Contact</a>
          <a class="nav-item nav-link" href="/download">Download</a>
            <a class="nav-item nav-link" href="https://github.com/QuackenbushLab/grand"><i class="fab fa-github fa-lg"></i></a>
        </div>
    </div>
  </nav>


<div class="container mb-5">
    <ol class="breadcrumb">
    <li class="breadcrumb-item"><a href="#networks">Networks</a></li>
    <li class="breadcrumb-item"><a href="#analysis">Analysis</a></li>
    <li class="breadcrumb-item"><a href="#tuto">API</a></li>
    <li class="breadcrumb-item"><a href="#credit">Image credit</a></li>
    </ol>
    <p><h3 id="networks" class="text-primary">Networks</h3>
    <h4>Reconstruction</h4>
    <div class="ml-3">
    The networks can be reconstructed using the tool specified in the tool field in the data table. For example, the LCL cell line network can be built using the PANDA tool as implemented in the package netZooM version 0.1. The other fields in the datatables refer to the arguments or data sources needed for the tool. For example, PANDA requires a TF-TF PPI network, gene expression samples, and TF motif binding site. 
    These data files along with their sources are provided to reconstruct the network. Since PANDA is a deterministic algorithm, reconstructing the network using the data priors and the same software release should yield the exact same network as the one available for direct download.
    </div>
    <h4>Download</h4>
    <div class="ml-3">The networks can be downloaded with two different properties.
    <h5>Format:</h5>
    <code>p:</code> stands for <code>.pairs</code> format. The network is written in a text or csv file with the following format:
    <pre>                node1    node2    edge_weight 
    <br>          e.g.,  A        B        1.0</pre>
    <br> or in the following format for multi-sample files:
    <pre>                node1:node2    edge_weight_sample_1   edge_weight_sample_n
    <br>          e.g.,  A:B            1.0                    2.0</pre>
    <br><code>c:</code> stands for <code>.csv</code> format. The bipartite network is saved as the weighted adjacency matrix W(TFs,Genes).
    <br><code>m:</code> stands for <code>.mat</code> format. The bipartite network is saved as the weighted adjacency matrix W(TFs,Genes) in MATLAB format.
    <h5>Edge weights:</h5>
    <code>o:</code> stands for original edge weights. These are the original edges weights computed by the <a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0064832">PANDA</a> algorithm. They usually vary between -20 and 20.
    A large value means a high likelihood of the existence of an edge between two nodes, a low negative value means a small probability of interaction between two given nodes.
    <br><code>t1:</code> stands for transformation 1. This is the transformation of original edge weights into a positive quantity. The transformation is detailed in <a href="https://www.sciencedirect.com/science/article/pii/S2211124717314183">Sonawane et al, Cell reports, 2017</a>.
    <br><code>t2:</code> stands for transformation 2. This is a transformation of the original edge weights into a quantity between 0 and 1 using a logistic regression of parameter 0.3.
    <br>Download options will be in <code>&lt;format&gt;&lt;edge weights&gt;</code>. For example <button type="button" class="btn btn-outline-primary">po</button> allows to download a network in <code>.pairs</code> format with original edge weights.</p>
    <div class="alert alert-warning alert-dismissible fade show" role="alert">
    <strong>Important!</strong> All the formats and edge weight transformations keeps the structure of the gene regulatory network as a complete bipartite graph of genes*TFs edges.
    <button type="button" class="close" data-dismiss="alert" aria-label="Close">
    <span aria-hidden="true">&times;</span>
    </button>
    </div>
    </div>
    <p><h3 id="analysis" class="text-primary">Analysis</h3>
    <div class="ml-3"><p align="justify"> The analysis section allows to analyse a set of TFs or genes in disease or small molecule category. It exploits the duality between TFs and genes
    in bipartite gene regulatory networks.</p>
    <p align="justify">The small molecule analysis section allows to find compounds that optimally reverse the gene targeting or the transcription factor activity patterns in the query set.</p>
    <p align="justify">Gene targeting refers the to the weighted in-degree of a given gene. Since PANDA is usually validated against Chip-Seq data, targeting can be interpreted as the binding profile of TFs for a given gene in a particular experiment. Transcription targeting activity refers to the weighted out-degree of a given transcription factor. This tool serves for hypothesis generation wheras compounds that reverse/aggravate the gene/TF targeting patterns of a given experiment are hypothetical candidates for experimental validation. </p>
    <p align="justify">The overlap score is equal to intersect(Input_Genes_Up,Drug_Genes_down) + intersect(Input_Genes_Down,Drug_Genes_Up) - intersect(Input_Genes_Up,Drugs_Genes_Up) - intersect(Input_Genes_Down,Drugs_Genes_Down). The cosine similarity compares the input vector to all the drugs. It is a signed measure that computes the direction of the vectors rather than their amplitude. In our case, we are interested in measuring the vectors of opposite directions, thus having negative cosine similarity.</p>
    </div>
    </p>
    <p><h3 id="tuto" class="text-primary">API</h3>
    <h4>Documentation</h4>
    <div class="ml-3">GRAND has an API that allows accessing the data programmatically to perform large-scale downloads. The full API documentation is available at <a href="https://grand.networkmedicine.org/redoc/">https://grand.networkmedicine.org/redoc/</a>.
    A web view of the API is also available for <a href="https://grand.networkmedicine.org/api/v1/cellapi/">cells</a>, <a href="https://grand.networkmedicine.org/api/v1/tissueapi/">tissues</a>, and <a href="https://grand.networkmedicine.org/api/v1/drugapi/">drug networks</a>.
    </div>
    <h4>Tutorial 1</h4>
    <div class="ml-3">You can access the database programmatically through the API, here we provide an example using the <a href="https://requests.kennethreitz.org/en/master/">requests</a> library of Python3. Additionally, you need to install the library <a href="https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html">awscli</a> to download networks through the command line.
    <pre><code>
    import requests
    import os      
    </code></pre>
    We perform a GET operation to the drugs database but we can also query cells and tissues. We need to make sure the returned status code is 200.
    <pre><code>
    response=requests.get('https://grand.networkmedicine.org/api/v1/drugapi/')
    
    if response.status_code == 200:
    	print('Success!')
    elif response.status_code == 404:
    	print('Not Found.')
    </code></pre>
    Since the API results are paginated for faster access, the previous command returns the first page with the first 50 drugs. To get the address of the next page, please use the following command:
    <pre><code>
    data['next']
    </code></pre>
    Then, we transform the result data into JSON strings for easier parsing.
    <pre><code>
    data=response.json()
    drugs=data['results']
    </code></pre>
    We can print the name of the first 50 drugs in the database among other attributes.
    <pre><code>
    for drug in drugs:
    	print(drug['drug'])
    </code></pre>
    We can also serially download the drug-induced gene regulatory network through awscli library. You can wrap this command into a for loop:
    <pre><code>
    os.system('aws s3 cp ' + drug['network'] + ' .')
    </code></pre>
    </p>
    </div>
    <h4>Tutorial 2</h4>
    <div class="ml-3">It is possible to rebuild the networks hosted in grand database using the metadata provided with each network. For example, we can rebuild the <a href='/cell'>LCL cell line network</a> using the provided prior data.
    A closer look at the LCL cell line entry shows that the network has been reconstructed using the tool PANDA, as implemented in <a href='http://netzoo.github.io'>netZooM</a> version 0.1. PANDA takes as input, a PPI network,
    gene expression data, and TF DNA binding motif data. These priors are provided along with the network. Since the network has been reconstructed in MATLAB, the following MATLAB tutorial goes through all the steps of downloading
    the priors through programmatic access and the reconstruction of the network: <a href='https://github.com/netZoo/netZooM/tree/devel/tutorials/grand_db'>Using GRAND database's API for reproducible network reconstruction</a>.</div>
    <p><h3 id="credit" class="text-primary">Image credit</h3>
    <div class="ml-3">Small molecule infographic: <a href="https://www.freepik.com/free-photos-vectors/background">Background vector created by freepik - www.freepik.com</a>
    <br>Tissues infographic: <a href="https://www.freepik.com/free-photos-vectors/heart">Heart vector created by freepik - www.freepik.com</a>
    <br>Cell lines infographic: <a href="https://www.freepik.com/free-photos-vectors/background">Background vector created by brgfx - www.freepik.com</a>
    <br>
    <br>
    <br></div>
    </p>
</div>

</div>
<!-- Footer -->
<footer class="page-footer font-small lighten-3 pt-4">
  <div class="container text-center text-md-left">
    <div class="row">
      <div class="col-md-6 col-lg-6">
        <h6 class="text-uppercase mb-2">Contact</h6>
        <p><a href="https://www.hsph.harvard.edu/biostatistics/">Department of Biostatistics</a>
        <br><a href="https://www.hsph.harvard.edu/">Harvard T.H. Chan School of Public Health</a></p>
      </div>
      <div class="col-md-5 col-lg-5">
        <!-- Social buttons -->
        <h6 class="text-uppercase mb-2">License</h6>
        <a class="nav-item nav-link" href="https://creativecommons.org/licenses/by-sa/4.0/"><i class="fab fa-creative-commons fa-2x"></i></a>
      </div>
      <div class="col-md-1 col-lg-1">
        <h6 class="text-uppercase mb-2">Share</h6>
        <!-- Facebook -->
        <a href="http://www.facebook.com/sharer/sharer.php?u=http://hsph.harvard.edu" type="button" class="btn-floating btn-fb">
          <i class="fab fa-facebook-square fa-2x"></i>
        </a>
        <!-- Twitter -->
        <a type="button" class="btn-floating btn-fb" href="https://twitter.com/intent/tweet?text=GRAND: Integrated gene regulatory network database of cell lines, tissues, and small molecules. https://grand.networkmedicine.org">
          <i class="fab fa-twitter-square fa-2x"></i>
        </a>
       <!-- Linkedin -->
        <a href="https://www.linkedin.com/sharing/share-offsite/?url=https://grand.networkmedicine.org" type="button" class="btn-floating btn-fb">
          <i class="fab fa-linkedin fa-2x"></i>
        </a>
        <a href="/about" type="button" class="btn-floating btn-fb">
          <i class="fas fa-envelope-square fa-2x"></i>
        </a>
      </div>
    </div>
  </div>
</footer>

<script src="https://kit.fontawesome.com/a076d05399.js"></script>
<script src="{% static 'js/jquery.slim.min.js' %}"></script>
<script src="{% static 'js/popper.min.js' %}"></script>
<script src="{% static 'js/bootstrap.min.js' %}"></script>
</body>
</html>

