{% load static %} 
{% include "header.html" %}

<div class="container mb-5">
  <div class="alert alert-secondary" role="alert">
    <nav aria-label="breadcrumb">
      <ol class="breadcrumb">
        <li class="breadcrumb-item"><a href="#networks" class="text-decoration-none">Networks</a></li>
        <li class="breadcrumb-item"><a href="#navigation" class="text-decoration-none">Navigation</a></li>
        <li class="breadcrumb-item"><a href="#analysis" class="text-decoration-none">Analysis</a></li>
        <li class="breadcrumb-item"><a href="#wiki" class="text-decoration-none">Wiki</a></li>
        <li class="breadcrumb-item"><a href="#api" class="text-decoration-none">API</a></li>
        <li class="breadcrumb-item"><a href="#credit" class="text-decoration-none">Image credit</a></li>
      </ol>
    </nav>
  </div>
    <div class="container">
      <div class="row">
        <div class="col-md-12 offset-md-2">
          <div id="highcontainer" style="width: 70%; height: 400px"></div>
        </div>
      </div>
    </div>
    GRAND works best on Google Chrome to take adavantage of the full network visualization features.
    <p><h3 id="networks" class="text-primary">Networks</h3>
    <h4>Reconstruction</h4>
    <div class="ml-3">
    The networks can be reconstructed using the tool specified in the tool field in the data table. For example, the LCL cell line network can be built using the PANDA tool as implemented in the package netZooM version 0.1. The other fields in the datatables refer to the arguments or data sources needed for the tool. For example, PANDA requires a TF-TF PPI network, gene expression samples, and TF motif binding site. 
    These data files along with their sources are provided to reconstruct the network. Since PANDA is a deterministic algorithm, reconstructing the network using the data priors and the same software release should yield the exact same network as the one available for visualization and download.
    </div>
    <h4>Download</h4>
    <div class="ml-3">The networks can be downloaded with two different properties.
    <h5>Format:</h5>
    <code>Edg:</code> stands for <code>edges</code> format. The network is written in a text or csv file with the following format:
    <pre>                node1    node2    edge_weight 
    <br>          e.g.,  A        B        1.0</pre>
    <br> or in the following format for multi-sample files:
    <pre>                node1:node2    edge_weight_sample_1   edge_weight_sample_n
    <br>          e.g.,  A:B            1.0                    2.0</pre>
    <br><code>Adj:</code> stands for <code>Adjacency</code> matrix format. The bipartite network is saved as the weighted adjacency matrix W(TFs,Genes).
    <br><code>gt:</code> stands for gene targeting. Gene targeting is the sum of weighted in-degrees in the network.
    <br><code>tt:</code> stands for TF targeting. TF targeting is the sum of weighted out-degrees in the network.
    <h5>Edge weights:</h5>
    The edges weights are computed by PANDA, PUMA, OTTER, and LIONESS algorithms. They usually vary between -20 and 20.
    A large value means a high likelihood of the existence of an edge between two nodes, a low negative value means a small probability of interaction between two given nodes. For DRAGON, edges weights are partial correlations between the nodes.
    <br>Download options are available in different formats. For example <button type="button" class="btn btn-outline-primary">Edg</button> allows to download a network in <code>edges</code> format with original edge weights.</p>
    <div class="container alert alert-warning alert-dismissible fade show" role="alert">
    <strong>Important!</strong> All gene regulatory networks are complete bipartite graphs of genes*TFs edges.
    <button type="button" class="btn-close" data-bs-dismiss="alert" aria-label="Close">
    </button>
    </div>
    </div>
    <p><h3 id="navigation" class="text-primary">Navigation</h3>
      <div class="row">
        <div class="col">
          <h4 id="navigation" class="text-primary">Small molecule drugs</h4>
          <img src="https://granddb.s3.us-east-2.amazonaws.com/gifs/drugs.gif" width="500" height="200">
        </div>
        <div class="col">
          <h4 id="navigation" class="text-primary">Cancer types</h4>
          <img src="https://granddb.s3.us-east-2.amazonaws.com/gifs/cancer.gif" width="500" height="200">
        </div>
      </div>
      <div class="row">
        <div class="col">
        <h4 id="navigation" class="text-primary">Tissues</h4>
        <img src="https://granddb.s3.us-east-2.amazonaws.com/gifs/tissues.gif" width="500" height="200">
      </div>
      <div class="col">
      <h4 id="navigation" class="text-primary">Cell lines</h4>
      <img src="https://granddb.s3.us-east-2.amazonaws.com/gifs/cell.gif" width="500" height="200">
      </div>
    </div>
    <p><h3 id="analysis" class="text-primary">Analysis</h3>
    <div class="ml-3"><p align="justify"> The analysis section allows to analyse a set of TFs or genes in disease or small molecule category. It exploits the duality between TFs and genes
    in bipartite gene regulatory networks.</p>
    <p align="justify">The small molecule analysis section allows to find compounds that optimally reverse the gene targeting or the transcription factor activity patterns in the query set.</p>
    <p align="justify">Gene targeting refers the to the weighted in-degree of a given gene. Since PANDA is usually validated against ChIP-seq data, targeting can be interpreted as the binding profile of TFs for a given gene in a particular experiment. Transcription targeting activity refers to the weighted out-degree of a given transcription factor. This tool serves for hypothesis generation wheras compounds that reverse/aggravate the gene/TF targeting patterns of a given experiment are hypothetical candidates for experimental validation. </p>
    <p align="justify">The overlap score is equal to card(intersect(Input_Genes_Up,Drug_Genes_down)) + card(intersect(Input_Genes_Down,Drug_Genes_Up))
     - card(intersect(Input_Genes_Up,Drugs_Genes_Up)) - card(intersect(Input_Genes_Down,Drugs_Genes_Down)). The cosine similarity compares the input vector to all the drugs. It is a signed measure that computes the direction of the vectors rather than their amplitude. In our case, we are interested in measuring the vectors of opposite directions, thus having negative cosine similarity.</p>
    </div>
    <div class="row">
      <div class="col">
      <h4 id="navigation" class="text-primary">Network view</h4>
      <img src="https://granddb.s3.us-east-2.amazonaws.com/gifs/network.gif" width="500" height="250">
    </div>
    <div class="col">
    <h4 id="navigation" class="text-primary">Targeting view</h4>
    <img src="https://granddb.s3.us-east-2.amazonaws.com/gifs/targeting.gif" width="500" height="250">
    </div>
  </div>
    </p>
    <p><h3 id="wiki" class="text-primary">Wiki</h3>
    <h4>PANDA</h4>
    PANDA is a method that allows the inference of a TF to gene bipartite gene regulatory network by integrating three data sources: 1) gene coexpression, 2) TF PPI network, and 3) TF motif network. Details and examples can be found <a href="https://netzoo.github.io/zooanimals/panda/" class="text-decoration-none" target="_blank">here</a>.
    PANDA was compared to <a href='https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0064832' class="text-decoration-none" target="_blank">four other inference methods<a> using ChIP-ChIP data in yeast in three conditions: TF knockout (KO), cell cycle (CC), and stress response (SR).
    <div class=" col-12">
      <canvas id="cellbar" style="width:100%; height:400px;"></canvas>
    </div>
    <h4>OTTER</h4>
    OTTER is an implementation of PANDA that uses convex optimization to infer the gene regulatory network. OTTER assumes that PPI and coexpression networks are projections of the regulatory network on the TF and Gene space and infers the network by optimzing graph matching between the three input networks. Details and examples can be found <a href="https://netzoo.github.io/zooanimals/otter/" class="text-decoration-none" target="_blank">here</a>.
    The accuracy of OTTER networks has been benchmarked against ChIP-seq data in <a href="https://grand.networkmedicine.org/cancers/Breast_cancer/" class="text-decoration-none" target="_blank">breast</a> cancer, <a href="https://grand.networkmedicine.org/cancers/Liver_cancer/" class="text-decoration-none" target="_blank">liver</a> cancer, and <a href="https://grand.networkmedicine.org/cancers/Cervix_cancer/" class="text-decoration-none" target="_blank">cervix</a> cancer.
    <h4>PUMA</h4>
    PUMA is an implementation of PANDA that infers miRNA to gene bipartite networks by integrating two data sources: 1) gene coexpression and 2) miRNA target networks. Details and examples can be found <a href="https://netzoo.github.io/zooanimals/puma/" class="text-decoration-none" target="_blank">here</a>.
    <h4>LIONESS</h4>
    LIONESS allows to estimate single-sample gene regulatory networks from one aggregate network such as PANDA, PUMA, or coexpression networks. Details and examples can be found <a href="https://netzoo.github.io/zooanimals/lioness/" class="text-decoration-none" target="_blank">here</a>.
    <h4>DRAGON</h4>
    DRAGON allows the simultaneous inference of multi-layer Gaussian Graphical Model (GGM) omic networks. DRAGON allows the estimation of partial correlations using a low number of sample by implementing covariance shrinkage. Details and examples can be found <a href="https://netzoo.github.io/zooanimals/dragon/" class="text-decoration-none" target="_blank">here</a>.
    DRAGON was compared to GGM for the accuracy to identify edges in single-cell multiomic dataset of simultaneously measured transcitptome and epitope (<a href="https://www.nature.com/articles/nmeth.4380" class="text-decoration-none" target="_blank">CITE-seq</a>) for 6 diffrent sample sizes.
    <div class=" col-12">
      <canvas id="dragonbar" style="width:100%; height:400px;"></canvas>
    </div>
    <h4>Gene/TF/miRNA targeting</h4>
    Targeting is a score for dircted bipartite networks. For source nodes (TFs/miRNA), targeting is the weighted outdegree, i.e., the sum of edge weights originating from the node. For target nodes (Genes), targeting is the weighted indegree, i.e., the sum of edge weights reaching the node. 
    This score has been detailed in <a href="https://arxiv.org/abs/2101.03985" class="text-decoration-none" target="_blank">Weighill et al. 2021</a>.
    <h4>GTEx</h4>
    GTEx is a project that collected samples from 38 normal, non-disased tissues in humans and measured gene expression across each tissues. More details can be found <a href="https://www.gtexportal.org/home/" class="text-decoration-none" target="_blank">here</a>.
    <h4>TCGA</h4>
    TCGA hosts gene expression data as well as other genomic information such as mutations, copy number variation and methylation for several cancer types in different tissues. More details can be found <a href="https://portal.gdc.cancer.gov/" class="text-decoration-none" target="_blank">here</a>.
    <h4>CCLE</h4>
    CCLE datasets characterize more than 1500 human cell lines that model several cancer types across a large array of omic layers including copy number variation, mutations, methylation, protein and metabolite levels, alternative splicing, and chromatin marks quantification. More details can be found <a href="https://portals.broadinstitute.org/ccle" class="text-decoration-none" target="_blank">here</a>.
    <h4>Connectivity Map</h4>
    The connectivity map measured the activity of more than 20,000 approved and experimental small molecule compounds in normal and cancerous cell lines. The expression of 987 genes was measured using the L1000 assay and the expression of close to 11,000 genes were inferred from the 1000 genes set. More details can be found <a href="https://clue.io" class="text-decoration-none" target="_blank">here</a>.
    <p><h3 id="api" class="text-primary">API</h3>
    <h4>Documentation</h4>
    <div class="ml-3">GRAND has an API that allows accessing and filtering the data programmatically to perform large-scale downloads. The full API documentation is available at <a class="text-decoration-none" target="_blank" href="https://grand.networkmedicine.org/redoc/">https://grand.networkmedicine.org/redoc/</a>.
    A web view of the API is also available for <a class="text-decoration-none" target="_blank" href="https://grand.networkmedicine.org/api/v1/cancerapi/">cancer type</a>, <a class="text-decoration-none" target="_blank" href="https://grand.networkmedicine.org/api/v1/geneapi/">genes</a>, <a class="text-decoration-none" target="_blank" href="https://grand.networkmedicine.org/api/v1/tissueapi/">tissues</a>, and <a class="text-decoration-none" target="_blank" href="https://grand.networkmedicine.org/api/v1/drugapi/">drug networks</a>.
    </div>
    <h4>Tutorial 1</h4>
    <div class="ml-3">You can access the database programmatically through the API, here we provide an example using the <a target="_blank" style="text-decoration:none" href="https://requests.kennethreitz.org/en/master/">requests</a> library of Python3. Additionally, you need to install the library <a target="_blank" style="text-decoration:none" href="https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html">awscli</a> to download networks through the command line.
    <pre><code>
    import requests
    import os      
    </code></pre>
    We perform a GET operation to access the drugs database but we can also query cancer type, genes, cells, and tissues. We need to make sure the returned status code is 200.
    <pre><code>
    response=requests.get('https://grand.networkmedicine.org/api/v1/drugapi/')
    </code>
    </pre>
    Filtering can be made server-side through querying the attributes of the network of interest. The full list of attributes is available at the <a target="_blank" style="text-decoration:none" href='https://grand.networkmedicine.org/redoc/'>documentation page</a>. For example, to query by drug name, we can use the following command:
    <pre>
    <code>
    responseFiltered=requests.get('https://grand.networkmedicine.org/api/v1/drugapi/?drug=1-phenylbiguanide')
    </code>
    </pre>
    To check that the data was collected correclty, we can do the following test.
    <pre>
    <code>
    if response.status_code == 200:
    	print('Success!')
    elif response.status_code == 404:
    	print('Not Found.')
    </code></pre>
    Since the API results are paginated for faster access, the previous command returns the first page with the first 50 drugs. To get the address of the next page, please use the following command:
    <pre><code>
    data['next']
    </code></pre>
    Then, we transform the result data into JSON strings for easier parsing.
    <pre><code>
    data=response.json()
    drugs=data['results']
    </code></pre>
    We can print the name of the first 50 drugs in the database among other attributes.
    <pre><code>
    for drug in drugs:
    	print(drug['drug'])
    </code></pre>
    We can also serially download the drug-induced gene regulatory network through awscli library. You can wrap this command into a for loop:
    <pre><code>
    os.system('curl -O ' + drug['network'] + ' .')
    </code></pre>
    </p>
    </div>
    <h4>Tutorial 2</h4>
    <div class="ml-3">It is possible to rebuild the networks hosted in grand database using the metadata provided with each network. For example, we can rebuild the <a target="_blank" style="text-decoration:none" href='/cell'>LCL cell line network</a> using the provided prior data.
    A closer look at the LCL cell line entry shows that the network has been reconstructed using the tool PANDA, as implemented in <a target="_blank" style="text-decoration:none" href='http://netzoo.github.io'>netZooM</a> version 0.1. PANDA takes as input, a PPI network,
    gene expression data, and TF DNA binding motif data. These priors are provided along with the network. Since the network has been reconstructed in MATLAB, the following MATLAB tutorial goes through all the steps of downloading
    the priors through programmatic access and the reconstruction of the network: <a target="_blank" style="text-decoration:none"  href='https://github.com/netZoo/netZooM/tree/devel/tutorials/grand_db'>Using GRAND database's API for reproducible network reconstruction</a>.</div>
    The code is provided using <button type="button" class="btn btn-outline-danger">Code</button> button. For networks generated using MATLAB, the .m script can be downloaded, however, for open-source languages (R and Python), the code is provided as Jupyter notebooks hosted on a cloud server called <a target="_blank" style="text-decoration:none" href="http://netbooks.networkmedicine.org">netbooks</a>.
    <p><h3 id="credit" class="text-primary">Image credit</h3>
    <div class="ml-3">Small molecule infographic: <a target="_blank" style="text-decoration:none" href="https://www.freepik.com/free-photos-vectors/background">Background vector created by freepik - www.freepik.com</a>
    <br>
    <br>
    <br></div>
    </p>
</div>

</div>
{% include "footer.html" %}

<script src="{% static 'js/highAnnoChart.js' %}"></script>
<script> var dataurl = "{% static 'js/annoChart.csv' %}";</script>

<script>
  var data2=JSON.parse("{{data|escapejs}}");
  var data2drag=JSON.parse("{{datadrag|escapejs}}");
</script>


<script src="{% static 'js/barChartPanda.js' %}"></script>

</body>
</html>

