{% load static %} 
{% include "header.html" %}
{% load widget_tweaks %}

{% if tfgeneseltar != 'nosel' %}
<div class="container alert alert-warning alert-dismissible fade show" role="alert">
  <button type="button" class="btn-close" data-bs-dismiss="alert" aria-label="Close">
  </button>

<h4 class="alert-heading">Submission stats</h4>
<p class="mb-1">
  {% if tfgeneseltar == 'by gene' %}
    {% if found == 'found' %}
    {{ ngenesfound }} genes found in network among {{ ngwas }} genes entered
    {% else %}
    Gene list not found, plotting network without gene filtering
    {% endif %}
  {% elif tfgeneseltar == 'by GO' %}
    {% if found == 'found' %}
    GO term found!
    {{ ngenesfound }} genes found among {{ ngwas }} genes in GO term
    {% else %}
    GO term not found, plotting network without GO filtering
    {% endif %}
  {% elif tfgeneseltar == 'by GWAS' %}
    {% if found == 'found' %}
    GWAS trait found!
    {{ ngenesfound }} genes found among {{ ngwas }} genes in GWAS trait
    {% else %}
    GWAS trait not found, plotting network without GWAS filtering
    {% endif %}
  {% elif tfgeneseltar == 'by tf' %}
    {% if found == 'found' %}
    {{ ngenesfound }} TFs/miRNA found in network among {{ ngwas }} TFs/miRNA entered
    {% else %}
    TF list not found, plotting network without TF filtering
    {% endif %}
  {% endif %}
        </p>
</div>
{% else %}
<div class="container alert alert-primary alert-dismissible fade show" role="alert">
  <button type="button" class="btn-close" data-bs-dismiss="alert" aria-label="Close">
  </button>

<p class="mb-1">
Please click on the submit button after selecting the parameters to retrieve the network from the server.
</p>
</div>
{% endif %}

<div class="modal hide fade" tabindex="-1" id="myModal">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
      </div>
      <div class="modal-body">
        <p>Please click on the submit button in the left panel to load the network from the server. The left panel allows to query and analyze the network using several parameters.</p>
        <p align="center"><img class="img-responsive" src="{% static 'images/submit-01.png' %}" height=200></p>
      </div>
      <div class="modal-footer">
        <button type="button" class="btn btn-secondary" data-bs-dismiss="modal">Close</button>
      </div>
    </div>
  </div>
</div>

<div class="modal hide fade" tabindex="-1" id="myModalTar">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
      </div>
      <div class="modal-body">
        <p>Please click on the submit button in the left panel to compute targeting scores. These scores can be further paramterized using the elements of the left panel. Next, it is possible to export the targeting scores to CLUEreg for drug repurposing or for TF enrichment analysis.</p>
        <p align="center"><img class="img-responsive" src="{% static 'images/submit_tar-01.png' %}" height=200></p>
      </div>
      <div class="modal-footer">
        <button type="button" class="btn btn-secondary" data-bs-dismiss="modal">Close</button>
      </div>
    </div>
  </div>
</div>

<div class='container pt-3 pb-4'>
  <div class="overlay"></div>
  <div class="spanner">
    <div class="loader"></div>
    <br><br><br>
    <p>Loading network, please wait.</p>
  </div>
    <div class="card text-center">
        <div class="card-header">
            <ul class="nav nav-pills card-header-pills" id="myTab" role="tablist">
                <li class="nav-item" role="presentation">
                  <a class="nav-link active btn" id="net-tab" data-bs-toggle="tab" data-bs-target="#network" role="tab" aria-controls="network" aria-selected="false">Differential network&nbsp;
                    <span class="translate-middle badge rounded-pill bg-light">
                    <img src="{% static 'images/question-solid.svg' %}" style="width: 10; height:15px" data-bs-toggle="modal" data-bs-target="#myModal">
                  </span></a>
                    <img src="{% static 'images/network.svg' %}" style="width: 115px; height:25px">
                </li>

                <li class="nav-item" role="presentation">
                    <a class="nav-link" id="tar-tab" data-bs-toggle="tab" data-bs-target="#tar" role="tab" aria-controls="tar" aria-selected="false">Differential targeting&nbsp;
                      <span class="translate-middle badge rounded-pill bg-light">
                        <img src="{% static 'images/question-solid.svg' %}" style="width: 10; height:15px" data-bs-toggle="modal" data-bs-target="#myModalTar">
                      </span></a>
                    </a>


                </li>
              </ul>
        </div>
    


        <!-- NETWORK -->
        <div class="tab-content" id="myTabContent">

            <div class="tab-pane fade active show" id="network" role="tabpanel" aria-labelledby="net-tab">

                <div class='container'>
                  <div class="row">
                    <div class="col-2 border border-secondary">
                        <br>
                        <button id="MSL-initialize-tour" type="button" class="btn btn-primary btn-sm">Guided tour</button>
                        <form action="{% url 'netcomp' slug %}" role="form" method="post">
                          {% csrf_token %}
                          <br>
                          <h5><span class="badge bg-secondary" data-step="1" data-intro="The first step starts by selecting two networks to compare and evaluate for diffrential targeting analysis. Network 1 can be a normal tissue and network 2 a cancer tissue, but the selection can be done to compare two cancer tissues or two normal tissues.">Network selection</span></h5>
                          <small><a href="#"
                            data-bs-toggle="popover"
                            data-bs-animation="false"
                            data-bs-placement="top"
                            data-bs-html="true"
                            data-bs-content="First network selection. Normal tissue networks were inferred from GTEx data and Cancer networks were inferred from TCGA. The tissue abbreviations can be found in <a target='_blank' href='https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/tcga-study-abbreviations'>TCGA web page</a>.">
                            Network 1: </a></small>
                            <small><div class="offset-2 col-8 offset-2"><div class="row">{{ netform.comp1 }} </div></div></small><br>    
                          <small><a href="#"
                            data-bs-toggle="popover"
                            data-bs-animation="false"
                            data-bs-placement="top"
                            data-bs-html="true"
                            data-bs-content="Second network selection. Normal tissue networks were inferred from GTEx data and Cancer networks were inferred from TCGA. The tissue abbreviations can be found in <a target='_blank' href='https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/tcga-study-abbreviations'>TCGA web page</a>.">
                            Network 2: </a></small> <br>
                          <small><div class="offset-2 col-8 offset-2"><div class="row">{{ netform.comp2 }} </div></div></small><br><br>
                          <h5><span  data-step="2" data-intro="Since these are genome-scale network, we can't plot all the edges on the browser, therefore, we need to select an edge subset." class="badge bg-secondary">Edges</span></h5>
                          <label for="myEdge" class="form-label">Number of edges</label>
                        <div>
                          <div class="offset-2 col-8 offset-2">
                            <div class='row'>
                          {{ netform.nedges }}
                            </div>
                          </div>
                          <p>Value: <span id="demo2"></span></p>
                          <div class='row'>
                            <div class='col-md-10 offset-md-1'>
                          <div class="form-group">
                            {% for radio in netform.topbottom %}
                            <div style="float: left;">
                              <small data-step="3" data-intro="After selecting the number of edges to plot, we can select wether these edges are the largest by edge weight or smallest.">{{ radio }} <a href="#"
                                data-bs-toggle="popover"
                                data-bs-animation="false"
                                data-bs-placement="top"
                                data-bs-html="true"
                                data-bs-content="Select the largest or the smallest edge weights in the network.">weights</a></small>
                              </div>
                            {% endfor %}
                          </div>
                        </div>
                      </div>
                          <div class='row'>
                            <div class='col-md-10 offset-md-1'>
                              <div style="float: left;">
                                {{ netform.brd }} <small data-step="4" data-intro="Selecting this option allows to plot edge thickness by edge weight, with red edges for negative values and green edges for positive values." >Weighted edges</small>
                              </div>
                              <div style="float: left;">
                                {{ netform.absval }} <small data-step="5" data-intro="The previous selection was done on signed values, however edge weights can be filtered by absolute values to plot the largest edges in both directions."><a href="#"
                                  data-bs-toggle="popover"
                                  data-bs-animation="false"
                                  data-bs-placement="top"
                                  data-bs-html="true"
                                  data-bs-content="Absolute value of edges">
                                Absolute value</a></small>
                              </div>
                              <div style="float: left;">
                                {{ netform.edgetargeting }} <small data-step="6" data-intro="This option allows to plot the edges that are supported by the presence of a TF motif in the promoter region of the target gene as dashed edges."><a href="#"
                                  data-bs-toggle="popover"
                                  data-bs-animation="false"
                                  data-bs-placement="top"
                                  data-bs-html="true"
                                  data-bs-content="Draws motif prior edges as dashed lines.">
                                Prior edges</a></small>
                              </div>
                            </div>
                          </div>
                        <br>
                          <h5><span class="badge bg-secondary">Node scaling</span></h5>
                          <div class='row'>
                            <div class='col-md-10 offset-md-1'>
                          <div style="float: left;">
                          {{ netform.dt.0 }} <small data-step="7" data-intro="Nodes can be scaled as well.">No scaling</small>
                          </div>
                          <div style="float: left;">
                          {{ netform.dt.1 }} <small data-step="8" data-intro="Their size can be proportional to their targeting scores.">Scale by <a href="#"
                            data-bs-toggle="popover"
                            data-bs-animation="false"
                            data-bs-placement="top"
                            data-bs-html="true"
                            data-bs-content="Scale node size by targeting score in the whole network. More about targeting score in <a href='https://grand.networkmedicine.org/help/#wiki'>GRAND help page</a>">
                          targeting</a></small>
                        </div>
                        <div style="float: left;">
                          {{ netform.dt.2 }} <small data-step="9" data-intro="Or by betweeness-centrality, which is a network metric computed on the selected subnetwork.">Scale by <a href="#"
                            data-bs-toggle="popover"
                            data-bs-animation="false"
                            data-bs-placement="top"
                            data-bs-html="true"
                            data-bs-content="Scale node size by betweness centrality measured in the selected subnetwork.">
                          betweeness</a></small>
                        </div>
                          </div>
                        </div>
                        <br>
                        <h5><span class="badge bg-secondary">Node selection</span></h5>
                        <div class='row'>
                          <div class='col-md-10 offset-md-1'>
                        <div style="float: left;">
                        {{ netform.tfgenesel.0 }} <small data-step="10" data-intro="We can also select only specific nodes to plot.">No selection</small>
                        </div>
                        <br>
                        <div style="float: left;">
                        {{ netform.tfgenesel.1 }} <small data-step="11" data-intro="The target nodes can be selected by their gene symbols or names.">by <a href="#"
                          data-bs-toggle="popover"
                          data-bs-animation="false"
                          data-bs-placement="top"
                          data-bs-html="true"
                          data-bs-content="Enter Gene names by ENSG ID, gene symbols, or mixed lists, with each gene in a new line">
                        gene</a></small>
                      </div>
                      <br>
                      {{ netform.geneform }}
                      <br>
                      <div style="float: left;">
                        {{ netform.tfgenesel.2 }} <small data-step="12" data-intro="The source nodes (TFs or miRNA) can be selected by gene symbols or names.">by <a href="#"
                          data-bs-toggle="popover"
                          data-bs-animation="false"
                          data-bs-placement="top"
                          data-bs-html="true"
                          data-bs-content="Enter TF names by ENSG ID, gene symbols, or mixed lists, with TF in a new line">
                        TF/miRNA</a></small>
                      </div>
                      <br>
                      {{ netform.tfform }}
                      <br>
                      <div style="float: left;">
                        {{ netform.tfgenesel.3 }} <small data-step="13" data-intro="Nodes belonging to specific pathways can be selected by entering the GO biological process for a pathway of interest.">by <a href="#"
                          data-bs-toggle="popover"
                          data-bs-animation="false"
                          data-bs-placement="top"
                          data-bs-html="true"
                          data-bs-content="Enter GO term ID for example GO:0060148. Please check <a href='https://grand.networkmedicine.org/pathways/' target='_blank'>the pathway page</a> for a full term list.">
                        GO</a></small>
                      </div>
                      <br>
                      {{ netform.goform }}
                      <br>
                      <div style="float: left;">
                        {{ netform.tfgenesel.4 }} <small data-step="14" data-intro="They can also be selected by GWAS trait from the GWAS catalog.">by <a href="#"
                          data-bs-toggle="popover"
                          data-bs-animation="false"
                          data-bs-placement="top"
                          data-bs-html="true"
                          data-bs-content="Enter GWAS catalog term for example AIDS progression. Please check <a href='https://grand.networkmedicine.org/gwascatalog/' target='_blank'>the GWAS catalog page</a> for a full term list.">
                        GWAS</a></small>
                      </div>
                      <br>
                      {{ netform.gwasform }}
                        </div>
                      </div>
                        <br>
                        <br>
                        <div class="form-actions button-box">
                          <div class="btn-group btn-group-sm" role="group" aria-label="Basic example">
                            <button type="button" class="btn btn-outline-primary" data-bs-toggle="modal" data-bs-target="#exampleModal2">
                            Info
                            </button>
                            <a href="#" data-step="16" data-intro="The subnetwork can be exported as a png file using this button." class="btn btn-outline-success" id="btn-download" download="subnetwork.png" >Save<br><small>as png</small></a>
                            <button  data-step="15" data-intro="Using all these parameters, the network will be first fetched from the server and then a subnetwork will be computed and plotted on the browser." type="submit"  name="submit" class="btn btn-primary" id="btnFetch">
                            <span id="mySpinner" class="d-none">
                            <span class="spinner-border spinner-grow-sm text-light" role="status" aria-hidden="true"></span>
                            </span>
                            Submit</button>
                            </div>

                        </div>
                        </form>
                        <br>
                        <br>
                        <p id="selection"></p>
                      </div>
                    </div>
                    
                    <div data-step="19" data-intro="Finally, the network will be plotted in this tab, hovering over the nodes gives additional information about the pathways and phenotypic traits." class="col-10 border border-secondary padding-0">
                      <div id="mynetwork" width=100% height=100%></div> 
                    </div>
                    <div class="col border border-secondary padding-0">
                      <div class="accordion" id="accordionExampleNN">
                        <div class="accordion-item">
                          <h2 class="accordion-header" id="headingOneNN">
                            <button data-step="18" data-intro="The subnetwork can be manipulated dyanmically without having to reload the parameters, using the following configuration panel."  class="accordion-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseOneNN" aria-expanded="true" aria-controls="collapseOneNN">
                              Configuration panel
                            </button>
                          </h2>
                          <div id="collapseOneNN" class="accordion-collapse collapse" aria-labelledby="headingOneNN" data-bs-parent="#accordionExampleNN">
                            <div class="accordion-body">
                              <div id="config"></div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                    <div class="card">
                      <ul class="list-group list-group-flush">
                        <div class="card-header">Network information</div>
                      <div class="card-body">
                        A set of networks in normal and cancer tissues have been generated using the same processing pipeline: The normal tissue networks are the ones that were generated in <a target="blank" style="text-decoration:none;" href="https://pubmed.ncbi.nlm.nih.gov/29069589/">Sonwane et al.</a> and available in the <a target="blank" style="text-decoration:none;" href="/tissues">tissues</a> section, and the cancer networks are available in the <a target="blank" style="text-decoration:none;"  href="/downloads">download</a> section. The processing pipeline can can be accessed in <a target="blank" style="text-decoration:none;" href="http://netbooks.networkmedicine.org">netbooks</a> under "Generating 26 cancer gene regulatory network using TCGA datasets". <br>To plot a differential network, first the first network (e.g., cancer) and the second network (e.g., normal tissue) can be selected from the dropdown menu and after clicking submit, the differential network is plotted by the specified parameters. Selecting 'weighted edges' option allows to plot edge thickness by the edge weight. Positive edge weights (network 2 > network 1) are plotted in green and negative edge weights (network 1 > network 2) are plotted in red.
                      </div>
                    </div>
                    <div class="card" data-step="17" data-intro="The network can be exported as an edge table using TFs as source and genes as targets, which can be directly used in Cytoscape.">
                    <div class="card-header">Edge weights table</div>
                    <div class="container col-10 mt-3 pb-3">
                      <table id="drugstable2" class="table table-hover" style="width: 100%">
                        <thead>
                          <tr>
                            <th scope="col">Source</th>
                            <th scope="col">Target</th>
                            <th scope="col">Edge weight</th>
                          </tr>
                        </thead>
                      </table>
                      </div>
                    </ul>
                  </div>
            </div>

        <!-- Targeting -->

        <div class="tab-pane fade show" id="tar" role="tabpanel" aria-labelledby="tar-tab">

          <div class='container'>
            <div class="row">
              <div class="col-2 border border-secondary">
                <div>
                  <form action="{% url 'difftaragg' slug %}" role="form" method="post">
                    {% csrf_token %}
                    <br>
                    <button id="MSL-initialize-tour3" type="button" class="btn btn-primary btn-sm">Guided tour</button>
                    <br><br>
                    <h5><span id="step1" class="badge bg-success">Step 1:<br> Targeting scores</span></h5>
                    <small><a href="#"
                      data-bs-toggle="popover"
                      data-bs-animation="false"
                      data-bs-placement="top"
                      data-bs-html="true"
                      data-bs-content="Normal tissue networks were inferred from GTEx data.">
                      <h5><span id="step0" class="badge bg-secondary">1. Network selection</span></h5>
                      Network 1: </a></small>
                      <small><div class="offset-2 col-8 offset-2"><div class="row">{{ tarform.comp11 }} </div></div></small><br>                           
                    <small><a href="#"
                      data-bs-toggle="popover"
                      data-bs-animation="false"
                      data-bs-placement="top"
                      data-bs-html="true"
                      data-bs-content="Cancer networks were inferred from TCGA. The tissue abbreviations can be found in <a target='_blank' href='https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/tcga-study-abbreviations'>TCGA web page</a>.">
                      Network 2: </a></small> <br>
                    <small><div class="offset-2 col-8 offset-2"><div class="row">{{ tarform.comp22 }} </div></div></small><br>
                    <h5><span class="badge bg-secondary">2. Genes</span></h5>
                    <div class="offset-2 col-8 offset-2">
                      <div class='row'>
                    <label for="myEdge" id="step2" class="form-label">Number of genes</label>
                    {{ tarform.nedgestar }}
                      </div>
                    </div>
                    <p>Value: <span id="demo2Tar"></span></p>
                    <div id="step3" class="form-group">
                      {% for radio in tarform.topbottomtar %}
                        <small>{{ radio }}</small>
                      {% endfor %}
                    </div>
                    <div class='row'>
                      <div class='col-md-10 offset-md-2'>
                        <div style="float: left;">
                          {{ tarform.absvaltar }} <small><a href="#"
                            data-bs-toggle="popover"
                            data-bs-animation="false"
                            data-bs-placement="top"
                            data-bs-html="true"
                            data-bs-content="Absolute value of targeting score">
                          Absolute value</a></small>
                        </div>
                      </div>
                    </div>
                    <br>
                    <h5><span class="badge bg-secondary">Gene selection</span></h5>
                    <div class='row'>
                      <div class='col-md-10 offset-md-2'>
                    <div style="float: left;">
                    {{ tarform.tfgeneseltar.0 }} <small>No selection</small>
                    </div>
                    <br>
                    <div style="float: left;">
                    {{ tarform.tfgeneseltar.1 }} <small id="step4">by <a href="#"
                      data-bs-toggle="popover"
                      data-bs-animation="false"
                      data-bs-placement="top"
                      data-bs-html="true"
                      data-bs-content="Enter Gene names by ENSG ID, gene symbols, or mixed lists, with each gene in a new line">
                    gene</a></small>
                  </div>
                  <br>
                  {{ tarform.geneformtar }}
                  <br>
                  <div style="float: left;">
                    {{ tarform.tfgeneseltar.2 }} <small id="step5">by <a href="#"
                      data-bs-toggle="popover"
                      data-bs-animation="false"
                      data-bs-placement="top"
                      data-bs-html="true"
                      data-bs-content="Enter GO term ID for example GO:0060148. Please check <a href='https://grand.networkmedicine.org/pathways/'>the pathway page</a> for a full term list.">
                    GO</a></small>
                  </div>
                  <br>
                  {{ tarform.goformtar }}
                  <br>
                  <div style="float: left;">
                    {{ tarform.tfgeneseltar.3 }} <small id="step6">by <a href="#"
                      data-bs-toggle="popover"
                      data-bs-animation="false"
                      data-bs-placement="top"
                      data-bs-html="true"
                      data-bs-content="Enter GWAS catalog term for example AIDS progression. Please check <a href='https://grand.networkmedicine.org/gwascatalog/'>the GWAS catalog page</a> for a full term list.">
                    GWAS</a></small>
                  </div>
                  <br>
                  {{ tarform.gwasformtar }}
                    </div>
                  </div>
                  <br>

                    <h5><span class="badge bg-secondary">3. TFs</span></h5>
                        <div class="offset-2 col-8 offset-2">
                          <div class='row'>
                            <label id="step7" for="myEdge" class="form-label">Number of TFs</label>
                        {{ tarform.nedgestartf }}
                          </div>
                        </div>
                        <p>Value: <span id="demo2Tartf"></span></p>
                        <div class="form-group">
                          {% for radio in tarform.topbottomtartf %}
                            <small>{{ radio }}</small>
                          {% endfor %}
                        </div>
                        <div class='row'>
                          <div class='col-md-10 offset-md-2'>
                            <div style="float: left;">
                              {{ tarform.absvaltartf }} <small><a href="#"
                                data-bs-toggle="popover"
                                data-bs-animation="false"
                                data-bs-placement="top"
                                data-bs-html="true"
                                data-bs-content="Absolute value of targeting score">
                              Absolute value</a></small>
                            </div>
                          </div>
                        </div>
                  <br>
                  <br>
                  <div class="form-actions button-box">
                    <div class="btn-group btn-group-sm" role="group" aria-label="Basic example">
                      <button type="button" class="btn btn-outline-primary" data-bs-toggle="modal" data-bs-target="#exampleModal">
                        Info
                      </button>
                    <button type="submit"  name="submit" class="btn btn-primary" id="btnFetchTar">
                      <span id="mySpinnerTar" class="d-none">
                      <span class="spinner-border spinner-grow-sm text-light" role="status" aria-hidden="true"></span>
                      </span>
                      Submit</button>
                      </div>
                  </div>
                </form>
                  <br>
                  <h5><span class="badge bg-success" id="step10">Step 2:<br> Analysis</span></h5>
                  <small>Analyze the submitted list of targeted genes/TFs for downstream analysis.</small>
                  <br>
                  <br>
                  <small>1. Enrichment analysis</small>
                  <div class="btn-group btn-group-sm" role="group" aria-label="Basic mixed styles example">
                    <button type="button" class="btn btn-outline-primary" data-bs-toggle="modal" data-bs-target="#exampleModalenrich">
                      Info
                    </button>
                    <a class="btn btn-outline-primary disabled" role="button">Enrich</a>
                    <a id="btnenrich" class="btn btn-outline-primary" href="{% url 'disease' %}" role="button">TFs</a>
                  </div>
                  <br><br>
                  <small>2. Drug repurposing</small>
                  <form action="{% url 'cluesubmit' 'redirect' %}" role="form" method="post">
                    {% csrf_token %}
                    {{ clueform.tfgeneselclue.0 }}<small id="step12">
                     By gene</small>
                    {{ clueform.tfgeneselclue.1 }}<small id="step13">
                     By TF</small>
                  <br><br>
                  <div class="btn-group btn-group-sm" role="group" aria-label="Basic mixed styles example">
                    <button type="button" class="btn btn-outline-primary" data-bs-toggle="modal" data-bs-target="#exampleModalclue">
                      Info
                    </button>
                    <button type="submit"  name="submit" class="btn btn-primary" id="btnFetchTarclue">
                      <span id="mySpinnerTar" class="d-none">
                      <span class="spinner-border spinner-grow-sm text-light" role="status" aria-hidden="true"></span>
                      </span>
                      CLUEreg</button>
                  </div>
                </form>
                  <br>
                  <br>
                </div>
              </div>
              <div id="step9" class='col-10 border border-secondary padding-0'> 
              <div class="row-4">
                <canvas id="tftardiv" style="width:100%; height:400px;"></canvas>
              </div>
              <div class="row-4">
                <canvas id="genetardiv" style="width:100%; height:400px;"></canvas>
              </div>
              </div>
            </div>
          </div>
          <div class="card">
            <ul class="list-group list-group-flush">
              <div class="card-header">Network information</div>
              <div class="card-body">
                A set of networks in normal and cancer tissues have been generated using the same processing pipeline: The normal tissue networks are the ones that were generated in <a target="blank" style="text-decoration:none;" href="https://pubmed.ncbi.nlm.nih.gov/29069589/">Sonwane et al.</a> and available in the <a target="blank" style="text-decoration:none;" href="/tissues">tissues</a> section, and the cancer networks are available in the <a target="blank" style="text-decoration:none;"  href="/downloads">downlaod</a> section. The processing pipeline can can be accessed in <a target="blank" style="text-decoration:none;" href="http://netbooks.networkmedicine.org">netbooks</a> under "Generating 26 cancer gene regulatory network using TCGA datasets". <br>To plot a differential network, first the first network (e.g., cancer) and the second network (e.g., normal tissue) can be selected from the dropdown menu and after clicking submit, the differential network is plotted by the specified parameters. Selecting 'weighted edges' option allows to plot edge thickness by the edge weight. Positive edge weights (network 2 > network 1) are plotted in green and negative edge weights (network 1 > network 2) are plotted in red.
            </div>
          </div>
          <div id="step15" class="card">
          <div class="card-header">Gene Targeting score</div>
          <div class="container col-10 mt-3 pb-3">
            <table id="drugstable3" class="table table-hover" style="width: 100%">
              <thead>
                <tr>
                  <th scope="col">Gene</th>
                  <th scope="col">Targeting score</th>
                </tr>
              </thead>
            </table>
            </div>
            <div class="card">
              <div class="card-header">TF Targeting score</div>
              <div class="container col-10 mt-3 pb-3">
            <table id="drugstable4" class="table table-hover" style="width: 100%">
              <thead>
                <tr>
                  <th scope="col">Regulator</th>
                  <th scope="col">Targeting score</th>
                </tr>
              </thead>
            </table>
            </div>
            </div>
          </ul>
        </div>
          </div>
          </div>
      </div>
    
      <div class="modal fade" id="exampleModal2" tabindex="-1" aria-labelledby="exampleModalLabel" aria-hidden="true">
        <div class="modal-dialog">
          <div class="modal-content">
            <div class="modal-header">
              <h5 class="modal-title" id="exampleModalLabel">Legend information</h5>
              <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
            </div>
            <div class="modal-body">
              <p style="text-align:left">Positive edge weights are colored in green, negative edge weights are colored in red. To export the network, please click on the save button to download a png image.</p>
            </div>
            <div class="modal-footer">
              <button type="button" class="btn btn-secondary" data-bs-dismiss="modal">Close</button>
            </div>
          </div>
        </div>
      </div>
      
<!-- Modal -->
<div class="modal fade" id="exampleModal" tabindex="-1" aria-labelledby="exampleModalLabel" aria-hidden="true">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <h5 class="modal-title" id="exampleModalLabel">Legend information</h5>
        <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
      </div>
      <div class="modal-body">
        <p style="text-align:left">Targeting refers to a bipartite network metric. For genes, targeting is the weighted indegree. For TFs, it represents the weighted outdegree. </p>
      </div>
      <div class="modal-footer">
        <button type="button" class="btn btn-secondary" data-bs-dismiss="modal">Close</button>
      </div>
    </div>
  </div>
</div>

<!-- Modal Enrich-->
<div class="modal fade" id="exampleModalenrich" tabindex="-1" aria-labelledby="exampleModalLabel" aria-hidden="true">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <h5 class="modal-title" id="exampleModalLabel">Information</h5>
        <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
      </div>
      <div class="modal-body">
        <p style="text-align:left">The enrich function submits the selected TFs for enrichment analysis. This feature is available only for TFs but not miRNA nor genes.</p> 
      </div>
      <div class="modal-footer">
        <button type="button" class="btn btn-secondary" data-bs-dismiss="modal">Close</button>
      </div>
    </div>
  </div>
</div>

<!-- Modal Enrich-->
<div class="modal fade" id="exampleModalclue" tabindex="-1" aria-labelledby="exampleModalLabel" aria-hidden="true">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <h5 class="modal-title" id="exampleModalLabel">Information</h5>
        <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
      </div>
      <div class="modal-body">
        <p style="text-align:left">The CLUEreg function submits the selected TFs/Genes in CLUEreg, which finds optimal compounds that reverse the TF/Gene profile. High-targeted genes/TFs are those that have a positive z-score and low-targeted genes are those with negative z-score. CLUEreg requires both high-targeted and low-targeted genes/TFs to be nonempty sets.</p>
      </div>
      <div class="modal-footer">
        <button type="button" class="btn btn-secondary" data-bs-dismiss="modal">Close</button>
      </div>
    </div>
  </div>
</div>

<!-- Modal same network-->
<div class="modal fade" id="exampleModalnet" tabindex="-1" aria-labelledby="exampleModalLabel" aria-hidden="true">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <h5 class="modal-title" id="exampleModalLabel">Information</h5>
        <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
      </div>
      <div class="modal-body">
        <p style="text-align:left">Please select 2 different networks.</p>
      </div>
      <div class="modal-footer">
        <button type="button" class="btn btn-secondary" data-bs-dismiss="modal">Close</button>
      </div>
    </div>
  </div>
</div>

</div><!-- End div for container-->


{% include "footer.html" %}

<script src="https://cdn.datatables.net/buttons/1.5.6/js/dataTables.buttons.min.js"></script>
<script src="https://cdn.datatables.net/buttons/1.5.6/js/buttons.flash.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jszip/3.1.3/jszip.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/pdfmake/0.1.53/pdfmake.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/pdfmake/0.1.53/vfs_fonts.js"></script>
<script src="https://cdn.datatables.net/buttons/1.5.6/js/buttons.html5.min.js"></script>
<script src="https://cdn.datatables.net/buttons/1.5.6/js/buttons.print.min.js"></script>


<script src="{% static 'js/tourenrich.js' %}"></script> 

<script type="text/javascript">
  var activetab="{{ activetab }}";
  console.log(activetab);
if(activetab == 'tar'){
  $("#tar").addClass('active');
  $("#network").removeClass('active');

  $("#tar-tab").addClass('active');
  $("#net-tab").removeClass('active');
}else{
  $("#network").addClass('active');
  $("#tar").removeClass('active');

  $("#net-tab").addClass('active');
  $("#tar-tab").removeClass('active');
}


</script>

<script>
  var isOpera = !!window.opera || navigator.userAgent.indexOf(' OPR/') >= 0;
    // Opera 8.0+ (UA detection to detect Blink/v8-powered Opera)
  var isFirefox = typeof InstallTrigger !== 'undefined';   // Firefox 1.0+
  var isSafari = Object.prototype.toString.call(window.HTMLElement).indexOf('Constructor') > 0;
      // At least Safari 3+: "[object HTMLElementConstructor]"
  var isChrome = !!window.chrome && !isOpera;              // Chrome 1+
  var isIE = /*@cc_on!@*/false || !!document.documentMode;
  // Edge 20+
  var isEdge = !isIE && !!window.StyleMedia;
  // Chrome 1+
  if (!isChrome){
    alert('Please switch to Google Chrome to enable network visualization.')
  }

  $(document).ready(function() {
    var edges = JSON.parse("{{edges|escapejs}}"); 
      //datatables
      $('#drugstable2').DataTable({
        "data": edges,
        "order": [[ 2, "desc" ]],
        "columns": [
           {"data": "sourcelabel"},
           {"data": "targetlabel"},
           {"data": "dispval"},
       ],
       buttons: [
                'copy', 'csv', 'pdf', 'print'
            ],
        "dom": "B" + "<'row'<'col-sm-12 col-md-6'l><'col-sm-12 col-md-5'f>>" +
               "<'row'<'col-sm-12'tr>>" +
               "<'row'<'col-sm-12 col-md-5'i><'col-sm-12 col-md-7'p>>",
      });
      
  
  } );
  $("#btnFetch").click(function() {
    $("div.spanner").addClass("show");
    $("div.overlay").addClass("show");
});
$("#btnFetchTar").click(function() {
    $("div.spanner").addClass("show");
    $("div.overlay").addClass("show");
});
</script>



<script>
  var slug="{{ slug }}";
  var slidertartf = document.getElementById("myEdgeTartf");
  var outputtartf = document.getElementById("demo2Tartf");
  outputtartf.innerHTML = slidertartf.value;
  
  slidertartf.oninput = function() {
    outputtartf.innerHTML = this.value;
  }
  var slidertar = document.getElementById("myEdgeTar");
  var outputtar = document.getElementById("demo2Tar");
  outputtar.innerHTML = slidertar.value;
  
  slidertar.oninput = function() {
    outputtar.innerHTML = this.value;
  }
  var slider = document.getElementById("myEdge");
  var output = document.getElementById("demo2");
  output.innerHTML = slider.value;
  
  slider.oninput = function() {
    output.innerHTML = this.value;
  }
  var nodes = JSON.parse("{{nodes|escapejs}}");  
  var edges = JSON.parse("{{edges|escapejs}}"); 
</script> 



<script>

  $(document).ready(function() {
      $("#btnFetch").click(function() {
          $("#mySpinner").removeClass("d-none")
      })
      $("#btnFetchTar").click(function() {
          $("#mySpinnerTar").removeClass("d-none")
      })
  });
  </script>
<script src="{% static 'js/networksagg.js' %}"></script>

<script>
      document.getElementById('id_comp1').onchange = function() {
        if ( document.getElementById('id_comp1').value==document.getElementById('id_comp2').value ){
          $('#exampleModalnet').modal('show');
          document.getElementById('btnFetch').disabled = true
        }else{
          document.getElementById('btnFetch').disabled = false
        }
      }
      document.getElementById('id_comp2').onchange = function() {
        if ( document.getElementById('id_comp1').value==document.getElementById('id_comp2').value ){
          $('#exampleModalnet').modal('show');
          document.getElementById('btnFetch').disabled = true
        }else{
          document.getElementById('btnFetch').disabled = false
        }
      }

      document.getElementById('id_comp11').onchange = function() {
        if ( document.getElementById('id_comp11').value==document.getElementById('id_comp22').value ){
          $('#exampleModalnet').modal('show');
          document.getElementById('btnFetchTar').disabled = true
        }else{
          document.getElementById('btnFetchTar').disabled = false
        }
      }
      document.getElementById('id_comp22').onchange = function() {
        if ( document.getElementById('id_comp11').value==document.getElementById('id_comp22').value ){
          $('#exampleModalnet').modal('show');
          document.getElementById('btnFetchTar').disabled = true
        }else{
          document.getElementById('btnFetchTar').disabled = false
        }
      }

      
  if(document.getElementById('absval')==null){
    if(slug.substr(0,3)=='ACH' || slug == 'mirnadragon' ){
      document.getElementById('id_edgetargeting').disabled = true; 
    }
    if(slug == 'mirnadragon' ){
      document.getElementById('id_tfgeneselclue_1').disabled = true; 
      var element = document.getElementById("btnenrich");
      element.classList.add("disabled");
    }
    document.getElementById('id_absval').onchange = function() {
        document.getElementById('id_brd').checked = this.checked;

    };
  }else if(document.getElementById('id_absval')==null){
    if(slug.substr(0,3)=='ACH' || slug == 'mirnadragon' ){
      document.getElementById('edgetargeting').disabled = true; 
      document.getElementById('tfgeneselclue_1').disabled = true; 
    }
    if(slug == 'mirnadragon' ){
      document.getElementById('tfgeneselclue_1').disabled = true; 
      var element = document.getElementById("btnenrich");
      element.classList.add("disabled");
    }
    document.getElementById('absval').onchange = function() {
        document.getElementById('brd').checked = this.checked;
    };
  }
</script>

<script>
var tftarscore = JSON.parse("{{tftarscore|escapejs}}");  
var genetarscore = JSON.parse("{{genetarscore|escapejs}}"); 
</script>


<script>
  $(document).ready(function() {
    var genetarscore = JSON.parse("{{genetarscore|escapejs}}"); 
      //datatables
      $('#drugstable3').DataTable({
        "data": genetarscore,
        "order": [[ 1, "desc" ]],
        "columns": [
           {"data": "index"},
           {"data": "tar"},
       ],
       buttons: [
                'copy', 'csv', 'pdf', 'print'
            ],
        "dom": "B" + "<'row'<'col-sm-12 col-md-6'l><'col-sm-12 col-md-5'f>>" +
               "<'row'<'col-sm-12'tr>>" +
               "<'row'<'col-sm-12 col-md-5'i><'col-sm-12 col-md-7'p>>",
      });
      
  
  } );
</script>

<script>
  $(document).ready(function() {
    var tftarscore = JSON.parse("{{tftarscore|escapejs}}");  
      //datatables
      $('#drugstable4').DataTable({
        "data": tftarscore,
        "order": [[ 1, "desc" ]],
        "columns": [
           {"data": "TF"},
           {"data": "tar"},
       ],
       buttons: [
                'copy', 'csv', 'pdf', 'print'
            ],
        "dom": "B" + "<'row'<'col-sm-12 col-md-6'l><'col-sm-12 col-md-5'f>>" +
               "<'row'<'col-sm-12'tr>>" +
               "<'row'<'col-sm-12 col-md-5'i><'col-sm-12 col-md-7'p>>",
      });
      
  
  } );
</script>


<script src="{% static 'js/bartar.js' %}"></script>



</script> 
</body>
</html>